File: backend/worker/handlers/title_overlay.py
from __future__ import annotations

from worker.handlers.base import BaseHandler


class TitleOverlayHandler(BaseHandler):
    async def execute(self, node_config, input_paths, output_path):
        video = input_paths["input"]
        text = str(node_config.get("text") or "")
        position = node_config.get("position", "top")
        start = float(node_config.get("start_time") or 0)
        duration = float(node_config.get("duration") or 3)
        font_size = int(node_config.get("font_size") or 72)
        safe_area = self.parse_bool_param(node_config.get("safe_area"), True)

        y_expr = {
            "top": "h*0.12" if safe_area else "h*0.06",
            "center": "(h-text_h)/2",
            "bottom": "h-text_h-h*0.18" if safe_area else "h-text_h-h*0.08",
        }.get(position, "h*0.12")
        escaped = text.replace("\\", "\\\\").replace(":", "\\:").replace("'", "\\'")
        enable = f"between(t,{start},{start + duration})"
        drawtext = (
            f"drawtext=text='{escaped}':fontcolor=white:fontsize={font_size}:"
            f"box=1:boxcolor=black@0.45:boxborderw=18:x=(w-text_w)/2:y={y_expr}:enable='{enable}'"
        )
        args = [
            "-i",
            video,
            "-vf",
            drawtext,
            *self.build_video_encode_args("libx264", preset="fast", crf=23),
            "-c:a",
            "aac",
            output_path,
        ]
        await self.run_ffmpeg(args)

File: internal/worker/handlers/title_overlay.go
package handlers

import (
	"context"
	"errors"

	vpffmpeg "github.com/Ctwqk/videoprocess/internal/worker/ffmpeg"
)

type TitleOverlayHandler struct {
	Runner vpffmpeg.Runner
}

func (h TitleOverlayHandler) NodeType() string {
	return "title_overlay"
}

func TitleOverlayArgs(inputPath string, outputPath string, config map[string]any) []string {
	text := stringValue(config["text"], "")
	position := stringValue(config["position"], "top")
	start := floatValue(config["start_time"], 0)
	duration := floatValue(config["duration"], 3)
	fontSize := intValue(config["font_size"], 72)
	safeArea := boolValue(config["safe_area"], true)

	yExpr := "h*0.12"
	switch position {
	case "top":
		if !safeArea {
			yExpr = "h*0.06"
		}
	case "center":
		yExpr = "(h-text_h)/2"
	case "bottom":
		if safeArea {
			yExpr = "h-text_h-h*0.18"
		} else {
			yExpr = "h-text_h-h*0.08"
		}
	}

	enable := "between(t," + formatFloat(start) + "," + formatFloat(start+duration) + ")"
	drawtext := "drawtext=text='" + escapeDrawText(text) + "':fontcolor=white:fontsize=" + intString(fontSize, 72) +
		":box=1:boxcolor=black@0.45:boxborderw=18:x=(w-text_w)/2:y=" + yExpr + ":enable='" + enable + "'"
	args := []string{"-i", inputPath, "-vf", drawtext}
	args = append(args, intermediateVideoEncodeArgs("libx264")...)
	args = append(args, "-c:a", "aac", outputPath)
	return args
}

func (h TitleOverlayHandler) Args(inputPath, outputPath string, config map[string]any) []string {
	return TitleOverlayArgs(inputPath, outputPath, config)
}

func (h TitleOverlayHandler) Execute(ctx context.Context, inputPaths map[string]string, outputPath string, config map[string]any) (map[string]any, error) {
	inputPath := inputPaths["input"]
	if inputPath == "" {
		return nil, errors.New("missing input path on input port")
	}
	if err := runFFmpeg(ctx, h.Runner, h.Args(inputPath, outputPath, config)); err != nil {
		return nil, err
	}
	return map[string]any{}, nil
}
